domain/cloud/state/state.go

```diff
@@ -73,6 +75,234 @@ func (st *State) ListClouds(ctx context.Context, filter *Filter) ([]cloud.Cloud,
 }
 
+// CloudDefaults provides the currently set cloud defaults for a cloud. If the
+// cloud has no defaults or the cloud does not exist a nil error is returned
+// with an empty defaults map.
+func (st *State) CloudDefaults(ctx context.Context, cloudName string) (map[string]string, error) {
+	defaults := map[string]string{}
+
+	db, err := st.DB()
+	if err != nil {
+		return defaults, errors.Trace(err)
+	}
+
+	stmt := `
+SELECT key, value
+FROM cloud_defaults
+INNER JOIN cloud
+ON cloud_defaults.cloud_uuid = cloud.uuid
+WHERE cloud.name = ?
+`
+
+	return defaults, db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		rows, err := tx.QueryContext(ctx, stmt, cloudName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		var key, value string
+		for rows.Next() {
+			if err := rows.Scan(&key, &value); err != nil {
+				return errors.Trace(stderrors.Join(err, rows.Close()))
+			}
+			defaults[key] = value
+		}
+		return nil
+	})
+}
+
+// UpdateCloudDefaults is responsible for updating default config values for a
+// cloud. This function will allow the addition and updating of attributes.
+// Attributes can also be removed by keys if they exist for the current cloud.
+func (st *State) UpdateCloudDefaults(
+	ctx context.Context,
+	cloudName string,
+	updateAttrs map[string]string,
+	removeAttrs []string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteBinds, deleteVals := database.SliceToPlaceholder(removeAttrs)
+	deleteStmt := fmt.Sprintf(`
+DELETE FROM cloud_defaults
+WHERE key IN (%s)
+AND cloud_uuid = (SELECT uuid
+                    FROM cloud
+                    WHERE name = ?)
+`[1:], deleteBinds)
+
+	upsertBinds, upsertVals := database.MapToMultiPlaceholder(updateAttrs)
+	upsertStmt := fmt.Sprintf(`
+CREATE TEMP TABLE cloud_default_dump (
+	key TEXT,
+    value TEXT
+);
+    
+INSERT INTO temp.cloud_default_dump VALUES %s;
+
+INSERT INTO cloud_defaults(cloud_uuid, key, value)
+SELECT (SELECT uuid FROM cloud WHERE name = ?) AS cloud_uuid,
+       key,
+       value
+FROM temp.cloud_default_dump
+-- The WHERE clause here is needed to help sqlite parser with ambiguity.
+WHERE true
+ON CONFLICT(cloud_uuid, key) DO UPDATE
+SET value = excluded.value
+WHERE cloud_uuid = excluded.cloud_uuid
+AND key = excluded.key;
+
+DROP TABLE temp.cloud_default_dump;
+`, upsertBinds)
+
+	return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		if len(deleteVals) != 0 {
+			_, err := tx.ExecContext(ctx, deleteStmt, append(deleteVals, cloudName)...)
+			if err != nil {
+				return fmt.Errorf("removing cloud default keys for %q: %w", cloudName, err)
+			}
+		}
+
+		if len(upsertVals) != 0 {
+			_, err := tx.ExecContext(ctx, upsertStmt, append(upsertVals, cloudName)...)
+			if sqliteErr, is := errors.AsType[sqlite3.Error](err); is && sqliteErr.ExtendedCode == sqlite3.ErrConstraintNotNull {
+				return fmt.Errorf("%w no cloud %q%w", errors.NotValid, cloudName, errors.Hide(err))
```

Comment:
won't this result in Error() being `not valid no cloud "mycloud"` which reads wrong

why don't we just stuck to the juju convention of `errors.NotValidf("missing cloud %q", cloudName)`
or
`errors.NewNotValid(err, "missing cloud %q", cloudName)`
or something like that

Reply:
I have rejigged to:

```
return fmt.Errorf("missing cloud %q %w%w", cloudName, errors.NotValid, errors.Hide(err))
```

This will embed both errors so they match for any further checking down the line but will hide the sql error in favor of the better error of: 'missing cloud "foo" not valid`.